New config parser, HCL support, multiple bind addrs

magiconair opened the pull request:
This patch implements a new config parser for the consul agent which
makes the following changes to the previous implementation:

 * add HCL support
 * all configuration fragments in tests and for default config are
   expressed as HCL fragments
 * HCL fragments can be provided on the command line so that they
   can eventually replace the command line flags.
 * HCL/JSON fragments are parsed into a temporary Config structure
   which can be merged using reflection (all values are pointers).
   The existing merge logic of overwrite for values and append
   for slices has been preserved.
 * A single builder process generates a typed runtime configuration
   for the agent.

The new implementation is more strict and fails in the builder process if no
valid runtime configuration can be generated. Therefore, additional
validations in other parts of the code should be removed.

The builder also pre-computes all required network addresses so that no
address/port magic should be required where the configuration is used and
should therefore be removed.

As a side effect the code also supports multiple bind addresses for HTTP,
HTTPS and DNS via go-sockaddr and static ip addresses.

Fixes #895
Fixes #2217 
Fixes #2825

magiconair commented:
There are still a handful of failing tests which I'll address tomorrow.

magiconair commented:
Still need to update all the docs ...

shilov commented:
👍  big thanks for adding HCL support + enabling multiple bind addresses

slackpad commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,4024 @@
+			},
+		},
+		{
+			desc:  "client addr, addresses and ports < 0",
```
Paused here.

slackpad replied:
Got to this spot - I'll continue from here...

magiconair commented on agent/agent.go:
```diff
@@ -362,22 +345,22 @@ func (a *Agent) listenAndServeDNS() error {
 			defer a.wgServers.Done()
-
-			err := s.ListenAndServe(p.Net, p.Addr, func() { notif <- p })
+			// todo(fs): does this also work for unix sockets?
```
The underlying DNS lib does not support Unix sockets. We can add this at a later stage.

slackpad replied:
Got it - yeah if DNS never could have supported it previously it's not important to get that in now (also seems like a fairly rare / low priority use case).

magiconair replied:
unless you want to test DNS on the socket without exposing it on a network interface. But lets add that later.

magiconair commented on agent/config/builder.go:
```diff
@@ -0,0 +1,1408 @@
+		Head:  []Source{DefaultSource()},
+	}
+
+	if b.stringVal(b.Flags.DeprecatedDatacenter) != "" && b.stringVal(b.Flags.Config.Datacenter) == "" {
```
This looks like a left over from moving code around.

magiconair replied:
It is a bit more complex than that but why do we want to carry on deprecated flags and fields to 1.0 in the first place? This looks like a good time to finally remove them.

magiconair commented on agent/config/config.go:
```diff
@@ -0,0 +1,381 @@
+	if strings.HasSuffix(name, ".hcl") {
+		return "hcl"
+	}
+	return "json"
```
The current config file logic is to assume any existing `-config-file` to be JSON and require extensions only for any file in a `-config-dir`. 

https://github.com/hashicorp/consul/blob/v0.9.3/agent/config.go#L2174-L2246

If we make this more strict then `consul agent -config-file consul.conf` will break, for example. The documentation also does not require an extension:

https://www.consul.io/docs/agent/options.html#configuration_files
https://www.consul.io/docs/agent/options.html#_config_file

It only mentions the extension in passing:

> When loading configuration, Consul loads the configuration from files and directories in lexical order. For example, configuration file basic_config.json will be processed before extra_config.json. 

I would also prefer to require a valid extension on all files.

magiconair commented on agent/config/config.go:
```diff
@@ -0,0 +1,381 @@
+	Watches                     []map[string]interface{} `json:"watches,omitempty" hcl:"watches"`
+
+	// non-user configurable values
+	ACLDisabledTTL             *string  `json:"acl_disabled_ttl,omitempty" hcl:"acl_disabled_ttl"`
```
They can be set from the outside but will be overwritten every time by the tail config fragment set in the `NewBuilder()`. 

https://github.com/hashicorp/consul/blob/refactor-config/agent/config/builder.go#L119

The default values are here:

https://github.com/hashicorp/consul/blob/refactor-config/agent/config/default.go#L104-L125

This is tested in the `TestFullConfig()` which has a custom non-user tail config:

https://github.com/hashicorp/consul/blob/refactor-config/agent/config/runtime_test.go#L3243-L3296

We could add an edge case test for each of these values to make it explicit.

magiconair commented on agent/config/config.go:
```diff
@@ -0,0 +1,381 @@
+	CleanupDeadServers      *bool   `json:"cleanup_dead_servers,omitempty" hcl:"cleanup_dead_servers"`
+	DisableUpgradeMigration *bool   `json:"disable_upgrade_migration,omitempty" hcl:"disable_upgrade_migration"`
+	LastContactThreshold    *string `json:"last_contact_threshold,omitempty" hcl:"last_contact_threshold"`
+	// todo(fs): do we need uint64 here? If yes, then I need to write a special parser b/c of JSON limit of 2^53-1 for ints
```
thought so. will fix

magiconair commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,4024 @@
+				rt.DataDir = dataDir
+			},
+		},
+		{
```
no, copy/paste left over when creating additional test cases.

magiconair replied:
I've removed it.

magiconair commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,4024 @@
+				rt.DataDir = dataDir
+			},
+		},
+		{ // todo(fs): shouldn't this be '-node-name'?
```
yeah, the fact that they are different is one of the issues.

slackpad commented:
Agree I was thinking the same and some of these are quite old - let's
remove them as long as we are clear in the upgrade guide and list them all.

On Sep 23, 2017 11:35 PM, "Frank Schröder" <notifications@github.com> wrote:

*@magiconair* commented on this pull request.
------------------------------

In agent/config/builder.go
<https://github.com/hashicorp/consul/pull/3480#discussion_r140648574>:

> +
+func NewBuilder(flags Flags) (*Builder, error) {
+	newSource := func(name string, v interface{}) Source {
+		b, err := json.MarshalIndent(v, "", "    ")
+		if err != nil {
+			panic(err)
+		}
+		return Source{Name: name, Format: "json", Data: string(b)}
+	}
+
+	b := &Builder{
+		Flags: flags,
+		Head:  []Source{DefaultSource()},
+	}
+
+	if b.stringVal(b.Flags.DeprecatedDatacenter) != "" &&
b.stringVal(b.Flags.Config.Datacenter) == "" {

It is a bit more complex than that but why do we want to carry on
deprecated flags and fields to 1.0 in the first place? This looks like a
good time to finally remove them.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/hashicorp/consul/pull/3480#discussion_r140648574>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AApG5VPOvrCujzugy2_jhMUcMRfURcMXks5slfgvgaJpZM4PblAP>
.

magiconair commented on agent/acl_test.go:
```diff
@@ -40,9 +52,9 @@ func (m *MockServer) GetPolicy(args *structs.ACLPolicyRequest, reply *structs.AC
-	cfg.ACLEnforceVersion8 = Bool(false)
-	a := NewTestAgent(t.Name(), cfg)
+	a := NewTestAgent(t.Name(), `
+ 		acl_enforce_version_8 = true
```
The default is set to a wrong key `acl_enforce_version8` (not the lacking final underscore) which may contribute to the test failing. 

I am working on a patch which detects invalid field names again and restores the previous functionality. However, since `encoding/json` and `hashicorp/hcl` decode embedded structs differently (`map[string]interface{}` vs `[]map[string]interface{}`). Therefore, just using `mapstructure` as before to detect unused fields won't work. This is a bug in HCL itself for which there is no easy fix AFAICT. Also, using the embedded JSON parser in `hcl` will also return `[]map[string]interface{}` which will trip up `mapstructure`. https://github.com/hashicorp/hcl/issues/209

Unless there is an option in mapstructure to do the mapping I think there are several viable approaches:

1. Patching all `[]map` to `map` and then use `mapstructure` for the mapping and invalid field detection. However, `watches` has `[]map[string]interface{}` so there would need to be an exception for that field and others in the future.

1. Build a custom invalid field detector which given a `map[string]interface{}` and struct type+tag name can detect which fields will not be mappable. 

1. Enhance the `hashicorp/hcl` decoder with an invalid field detector and do the same thing with `encoding/json` by first forking it and then adding that to the custom code until https://github.com/golang/go/issues/15314 is added to the stdlib.

1. Add that capability to `mapstructure`.

For the 1.0 milestone, the first option seems to be the one with the least amount of work so I'll try that first.

magiconair commented:
I've pushed a change which removes all deprecated fields and flags and keeps a list in `agent/config/doc.go` The last remaining thing is `recursor` since we have both `recursor` and `recursors` and usually merge the single recursor into the list.

slackpad commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,3562 @@
+			patch: func(rt *RuntimeConfig) {
+				rt.Bootstrap = true
+				rt.ServerMode = true
+				rt.LeaveOnTerm = false
```
In client mode the `LeaveOnTerm` and `SkipLeaveOnInt` values should be the opposite of server mode, so `true` and `false`, respectively. We should add a case that covers that behavior.

magiconair replied:
done.

slackpad commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,3562 @@
+			flags: []string{
+				`-data-dir=runtime_test.go`,
+			},
+			patch: func(rt *RuntimeConfig) {
```
Do we need a `patch` section when it's checking for an error?

slackpad commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,3562 @@
+			hcl:  []string{`dns_config = { udp_answer_limit = 0 }`},
+			err:  "dns_config.udp_answer_limit cannot be 0. Must be positive",
+		},
+		{
```
Duplicate of previous test.

magiconair replied:
removed

slackpad commented on agent/config/runtime_test.go:
```diff
@@ -0,0 +1,3562 @@
+				"dogstatsd_addr": "0wSndumK",
+				"dogstatsd_tags": [ "3N81zSUB","Xtj8AnXZ" ],
+				"filter_default": true,
+				"prefix_fi
```
Seems like we can eventually divorce the runtime config from the structure with both check and checks.

slackpad commented on agent/dns_test.go:
```diff
@@ -2913,7 +2858,7 @@ func TestDNS_ServiceLookup_AnswerLimits(t *testing.T) {
 		expectedANYQueryID  int
 	}{
-		{"0", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		//{"0", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
```
Curious what happened here? Does this fail to validate then we can just delete it.

magiconair replied:
The new code did not allow `dns_config.udp_answer_limit` to be zero which made this test fail. I guess we commented this out while getting things to work. I've restored the previous behavior.

slackpad commented on agent/node-id:
```diff
@@ -0,0 +1 @@
+9e8edccb-1f67-0ef6-a36f-e5121d91d8fb
```
This file should be deleted (or does a test use this)?

slackpad commented on agent/segment_stub.go:
```diff
@@ -2,18 +2,14 @@
-
-	return nil
-}
+// done(fs): func ValidateSegments(conf *Config) error {
```
Did this get moved elsewhere? There's an enterprise version of this file that's in the enterprise fork, so we had this out here so we can swap it with build tags.

magiconair replied:
I had incorporated this into the normal validation. However, to support different types of validation for enterprise I've split up the tests so that we can overload them. I'll send you the `ent` fragments you can add.

slackpad commented on agent/testagent_test.go:
```diff
@@ -0,0 +1,27 @@
+
+func TestDefaultConfig(t *testing.T) {
+	for i := 0; i < 500; i++ {
+		t.Run("bla", func(t *testing.T) {
```
`bla` might as well be an empty string

slackpad commented on ipaddr/detect.go:
```diff
@@ -0,0 +1,138 @@
+
+// GetPrivateIPv4 returns the list of private network IPv4 addresses on
+// all active interfaces.
+func GetPrivateIPv4() ([]*net.IPAddr, error) {
```
We could probably remove these helpers and change the default config to go-sockaddr templates that accomplishes this as well.

magiconair replied:
That makes mocking the address resolution more difficult. Let me think about a solution.

slackpad commented on vendor/github.com/hashicorp/go-sockaddr/ipv4addr.go:
```diff
@@ -58,7 +58,8 @@ func NewIPv4Addr(ipv4Str string) (IPv4Addr, error) {
 	// particular, clients with the Barracuda VPN client will see something like:
 	// `192.168.3.51/00ffffff` as their IP address.
-	trailingHexNetmaskRe := trailingHexNetmaskRE.Copy()
+	// trailingHexNetmaskRe := trailingHexNetmaskRE.Copy()
```
Is this a monkey patch? Why was this needed?

magiconair replied:
Yes, for some reason this creates a data race which made tests fail randomly at high concurrency. I still need to understand why that is.

slackpad commented:
I'll try to sort the Raft multiplier stuff.

slackpad commented:
@magiconair I'm thinking we can get rid of `PerformanceRaftMultiplier` from the runtime struct and do this scaling at the end of `Build()`, like we do for other derived parameters:

```
         base.RaftConfig.HeartbeatTimeout = a.config.ConsulRaftHeartbeatTimeout * time.Duration(a.config.PerformanceRaftMultiplier)
         base.RaftConfig.LeaderLeaseTimeout = a.config.ConsulRaftLeaderLeaseTimeout * time.Duration(a.config.PerformanceRaftMultiplier)
         base.RaftConfig.ElectionTimeout = a.config.ConsulRaftElectionTimeout * time.Duration(a.config.PerformanceRaftMultiplier)
```

slackpad commented:
@magiconair Heads up I pushed my small stuff from the -slackpad branch up onto yours, so please pull that.

magiconair commented:
Will do and will go through the comments. I have something working for the invalid fields. Not pretty but it should do for the beta. I'll go through the rest of the comments and fix/respond

magiconair commented on agent/agent_test.go:
```diff
@@ -167,13 +101,20 @@ func TestAgent_TokenStore(t *testing.T) {
 }
 
 func TestAgent_CheckPerformanceSettings(t *testing.T) {
+	t.Skip("this test fails because of the change in agent.consulConfig(). We need to decide what to do")
```
I've followed your suggestion and dropped the `PerformanceRaftMultiplier` parameter and instead compute the right values during the build step.

magiconair commented on agent/config/runtime.go:
```diff
@@ -0,0 +1,231 @@
+	return tc.IncomingTLSConfig()
+}
+
+func (c *RuntimeConfig) Sanitized() RuntimeConfig {
```
Added a test and implemented it reflectively. The expected test result is static so the test should fail once a new field with token, key or secret in the name is added. `gofuzz` failed on me with some error but I don't think it is necessary with this approach.

magiconair replied:
I've used different values for the WAN retry join case in a later commit.

magiconair commented:
I've rebased the branch onto master and added the code for detecting invalid fields. I've immediately found two tests which were broken so this is very useful. The approach is a bit hacky but I've added a long description of why we're doing this. If I can come up with a better approach then I'll try over the next days. 

The open issues are:

* drop `recursor` in favor of `recursors`
* look at the commented out section int `TestDNS_ServiceLookup_AnswerLimits`
* document/follow up the monkey patches for `hashicorp/go-sockaddr` and `hashicorp/hcl`
* require `.json` or `.hcl` extension for all config files. (need input/decision)

@slackpad you need to incorporate the enterprise test fragments for the segment code I've sent you.

slackpad commented:
That's fair and the mock is very nice for testing. We can keep this as is
going in to the beta.

On Sep 25, 2017 12:04 AM, "Frank Schröder" <notifications@github.com> wrote:

*@magiconair* commented on this pull request.
------------------------------

In ipaddr/detect.go
<https://github.com/hashicorp/consul/pull/3480#discussion_r140701179>:

> @@ -0,0 +1,138 @@
+package ipaddr
+
+import (
+	"fmt"
+	"net"
+)
+
+// GetPrivateIPv4 returns the list of private network IPv4 addresses on
+// all active interfaces.
+func GetPrivateIPv4() ([]*net.IPAddr, error) {

That makes mocking the address resolution more difficult. Let me think
about a solution.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/hashicorp/consul/pull/3480#discussion_r140701179>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AApG5bJGGhKzzQ0k7dFOpLYFoton9Y3lks5sl1B1gaJpZM4PblAP>
.

magiconair commented:
We also need to decide where to put the documentation that was part of `agent/config.Config`. We can either restore it to the `Config` struct, add it to the `RuntimeConfig` or omit it from code altogether and put it in the public documentation. Opinions?

magiconair commented:
F I N A L L Y !
